flow/record/base.py

```diff
@@ -338,243 +335,308 @@ class RecordFieldSet(list):
+        }
 
         Returns:
-            OrderedDict
-
+            Mapping of all Record fields
         """
-        required_fields = OrderedDict([(k, RecordField(k, v)) for k, v in RESERVED_FIELDS.items()])
-        return required_fields
+        if self._all_fields is None:
+            self._all_fields = self.fields.copy()
+            self._all_fields.update(self.get_required_fields())
+        return self._all_fields
 
-    def get_all_fields(self):
-        """
-        Get all fields including required meta fields.
+    def getfields(self, typename: str) -> RecordFieldSet:
+        """Get fields of a given type.
 
-        Returns:
-            OrderedDict
+        Args:
+            typename: The typename of the fields to return. eg: "string" or "datetime"
 
+        Returns:
+            RecordFieldSet of fields with the given typename
         """
-        required_fields = self.get_required_fields()
-        fields = self.fields.copy()
-        fields.update(required_fields)
-        return fields
-
-    def getfields(self, typename):
         if isinstance(typename, DynamicFieldtypeModule):
             name = typename.gettypename()
         else:
             name = typename
 
         return RecordFieldSet(field for field in self.fields.values() if field.typename == name)
 
-    def __call__(self, *args, **kwargs):
+    def __call__(self, *args, **kwargs) -> Record:
+        """Create a new Record initialized with `args` and `kwargs`."""
         return self.recordType(*args, **kwargs)
 
-    def init_from_dict(self, rdict, raise_unknown=False):
+    def init_from_dict(self, rdict: Dict[str, Any], raise_unknown=False) -> Record:
         """Create a new Record initialized with key, value pairs from `rdict`.
 
         If `raise_unknown=True` then fields on `rdict` that are unknown to this
         RecordDescriptor will raise a TypeError exception due to initializing
         with unknown keyword arguments. (default: False)
 
         Returns:
-            Record
-
+            Record with data from `rdict`
         """
 
         if not raise_unknown:
             rdict = {k: v for k, v in rdict.items() if k in self.recordType.__slots__}
         return self.recordType(**rdict)
 
-    def init_from_record(self, record, raise_unknown=False):
+    def init_from_record(self, record: Record, raise_unknown=False) -> Record:
         """Create a new Record initialized with data from another `record`.
 
         If `raise_unknown=True` then fields on `record` that are unknown to this
         RecordDescriptor will raise a TypeError exception due to initializing
         with unknown keyword arguments. (default: False)
 
         Returns:
-            Record
-
+            Record with data from `record`
         """
         return self.init_from_dict(record._asdict(), raise_unknown=raise_unknown)
 
-    def extend(self, fields):
+    def extend(self, fields: Sequence[Tuple[str, str]]) -> "RecordDescriptor":
```

Comment:
You could `from __future__ import annotations` to allow using `RecordDescriptor` here.

Reply:
I'll check if this is compatible with Python 3.7.

Reply:
Fixed in 4609670